File: Two_Pointers/q259-3sum-smaller.py
# 给定一个长度为 n 的整数数组和一个目标值 target，寻找能够使条件 nums[i] + nums[j] + nums[k] < target 成立的三
# 元组 i, j, k 个数（0 <= i < j < k < n）。
#
#  示例：
#
#  输入: nums = [-2,0,1,3], target = 2
# 输出: 2
# 解释: 因为一共有两个三元组满足累加和小于 2:
#      [-2,0,1]
#      [-2,0,3]
#
#
#  进阶：是否能在 O(n2) 的时间复杂度内解决？
#  Related Topics 数组 双指针
#  👍 52 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def threeSumSmaller(self, nums: List[int], target: int) -> int:
        if not nums or target is None: return 0
        nums.sort()
        cnt, n = 0, len(nums)
        for i in range(n - 2):  # ∵ rt_min = i+2 < n
            left, right = i + 1, n - 1
            while left < right:
                tmp_sum = nums[i] + nums[left] + nums[right]
                if tmp_sum >= target:
                    right -= 1
                else:
                    cnt += right - left
                    left += 1
        return cnt

# leetcode submit region end(Prohibit modification and deletion)

File: Two_Pointers/q259_3sum_smaller.java
package Two_Pointers;

import java.util.Arrays;

public class q259_3sum_smaller {
    public int threeSumSmaller(int[] nums, int target) {
        int n = nums.length;
        Arrays.sort(nums);
        int cnt = 0;
        // 法1：固定最大边
        for (int k = 2; k < n; k++) {
            int sum = target - nums[k]; // [i]+[j]<sum
            int i = 0, j = k-1;
            // System.out.println(i + ", " + j);
            while (i < j) {
                // -2(i), -2, -2, 0, 0(j), 3(k)
                if(nums[i] + nums[j] < sum) {
                    // ∵此后i需右移 ∴需要一次性将固定k&【i】的所有对数求出！
                    cnt += j-(i+1)+1; // 对数=[k]&【i】&[i+1, j]
                    i++; // ∑↑
                } else j--;// 不符合要求，∑↓
            }
        }

        // // 法2：固定最小边
        // for (int i = 0; i < n-2; i++) {
        //     int lf = i+1, rt = n-1;
        //     int sum = target - nums[i]; // [lf]+[rt]<sum
        //     // -2(i), -2(lf), -2, 0, 0, 3(rt)
        //     while (lf < rt) {
        //         if(nums[lf] + nums[rt] < sum) {
        //             cnt += rt - (lf+1) + 1; // 固定i&lf时の对数=[i]&【lf】&[lf+1, rt]
        //             lf++;  // ∑↑
        //         } else rt--;// 不符合要求，∑↓
        //     }
        // }
        return cnt;
    }
}
